lib/packwerk/application_validator.rb

```diff
@@ -134,6 +134,16 @@ def check_package_manifest_syntax
           end
         end
 
+        if hash.key?("public_path")
+          unless hash["public_path"].is_a?(String)
+            errors << "'public_path' option must be a string in #{f.inspect}: #{hash['public_path'].inspect}"
+          end
+
+          unless hash["public_path"].try(:last) == "/"
```

Comment:
`last` doesn't look like a core string method (https://ruby-doc.org/core-2.6/String.html) so, depending on the environment this runs in, this could always raise a `NoMethodError`. Best to index by -1 or `.chars.last`.

That being said, I'm thinking we should drop this anyways, since it doesn't guarantee a directory on disk. If the slash is necessary elsewhere in packwerk (I think so, because of [this](https://github.com/Shopify/packwerk/blob/da99ef83da6ba9fd0879c85501c6bd7ed4fa9ec0/lib/packwerk/package.rb#L40)), we should ensure it's there (perhaps in `Package#initialize`). I think this provides a slightly better user experience.

What do you think?

Reply:
Ah, good spot on the `#last`! Thanks for the review 🤓 

As for the validation... I assumed that the code I amended in application_validator would run on startup or something, is that not the case? And it's just run when the validate command is used via the CLI?

If it's not run outside of that, then I agree it would be nice to fail loudly, perhaps something like:

```
    def validate_public_path!
      return unless @config["public_path"]

      raise PublicPathMustEndInForwardSlash unless @config["public_path"].chars.last == "/"
    end
```

In package.rb, which is then called from the initializer?

I presume if we go with this, we'd also want to keep the validation in the application_validator too so that if people run that without the slash, it also intuitively lets them know?

Reply:
> As for the validation... I assumed that the code I amended in application_validator would run on startup or something, is that not the case? And it's just run when the validate command is used via the CLI?

You got it! This only gets executed when one runs `packwerk validate`. Just to explain why there's a separation here: `packwerk validate` uses Rails application state to help validate the configuration. `packwerk check` and `packwerk update` don't need to run with application state, which means they boot up and run much more quickly.

> In package.rb, which is then called from the initializer?

As I mentioned, I think it may be best to not even check for the slash but simply add it if it's missing. Imagine a scenario where it was configured to be `my_public`, but `my_public` was either a file or not the right directory.

First, we'd fail loudly saying you need to add a `/`, after which `packwerk check` would likely fail because the `start_with?` check (linked in my previous comment) would always fail. Once we correct that mistake, we'd likely have many privacy violations. It may be misleading to the end user to think that the validation is protecting them in case of mistake.

That being said, I can't think of a _great_ solution here. We _could_ check for the directory's existence (under the package path) when validation happens. The problem there is that maybe it's configured and there's no public directory at all? We could suggest, in the validator, to create the directory (possibly adding an empty file in there to ensure it is made available by your VCS). Every solution I can think of has pros and cons 🙈 

I'll poke someone on my team to chime in too. It looks like there's no perfect solution, so let's aim for at least a couple of people to agree on some approach, and go with that 🙂

Reply:
Personally I would say they are two different issues (and thus _probably_ two PRs). I think adding the slash in automatically makes sense, because before it was handled by the fact it was hard-coded, so we then have the same behavior with the custom paths.

Then the secondary quality of life change is around making sure that public folder exists, as realistically I imagine it's currently possible for people to not create the `app/public/` folder and it would have the same issue as if a custom path was used and that folder didn't exist (unless I am missing something, which is possible as I am new to the codebase 😅 ).

Reply:
I've pushed a change to automatically add the slash if it's missing, and removed the `/` handling in the validator (as there is no real need to loudly fail now we're handling it)

Reply:
I think it’s okay to configure a public folder path that doesn’t yet exist on disk. It’s a little weird but shouldn’t be treated as any more “wrong” than not having an `app/public` folder, which is fine.

There’s a deeper problem here which is that we shouldn’t really be manipulating paths with low-level string operations like concatenation and `#start_with?`, but that’s outside of the scope of this PR. I think the implementation looks good!